File: apps/web-e2e/src/support/fake-api.ts
import type { Page, Route } from '@playwright/test';

/**
 * A small in-memory stand-in for the API, wired through Playwright's route
 * interception.
 *
 * The point is to exercise the *customer journey* — catalogue, cart,
 * pickup slots, totals, order status — deterministically, in CI, without a
 * Postgres and a Redis. It is not a contract test: the shapes here are
 * hand-kept in step with the DTOs, and a change there will show up as a
 * failing journey rather than a silent pass.
 *
 * It carries real state between calls, so adding to the cart genuinely
 * changes the total the checkout renders. A pile of static fixtures would
 * pass while the app quietly failed to wire two screens together.
 */

export const STORE = {
  id: 'store-1',
  brandId: 'brand-1',
  slug: 'dubai-marina',
  name: 'takeAway Marina',
  addressLine: 'Marina Walk',
  city: 'Dubai',
  country: 'AE',
  latitude: 25.078,
  longitude: 55.141,
  status: 'OPEN',
  fulfillmentTypes: ['TAKEAWAY'],
  pickupPointType: 'SHELF',
  busyMeter: 20,
  currentEtaSeconds: 420,
  // 5% VAT already inside the price, as in the UAE.
  taxRateBps: 500,
  taxIncludedInPrice: true,
  currency: 'AED',
  heroImageUrl: null,
  distanceMeters: null,
  timezone: 'Asia/Dubai',
  phone: null,
  email: null,
  minOrderCents: 0,
  galleryUrls: [],
  workingHours: [],
  brand: { id: 'brand-1', slug: 'takeaway', name: 'takeAway', logoUrl: null, themeOverrides: null },
};

/**
 * Shaped to `ProductDetail` in @takeaway/shared-types. Every array field is
 * present even when empty: the product template reads `.length` on
 * allergens and modifiers, and an omitted key throws during change
 * detection rather than rendering an empty section.
 */
export const PRODUCT = {
  id: 'product-1',
  categoryId: 'cat-1',
  slug: 'flat-white',
  name: 'Flat White',
  description: 'Double ristretto, steamed milk.',
  basePriceCents: 1800,
  prepTimeSeconds: 120,
  caffeineLevel: 2,
  calories: 120,
  proteinsGrams: 6,
  fatsGrams: 5,
  carbsGrams: 9,
  allergens: [],
  dietTags: [],
  imageUrls: [],
  sortOrder: 0,
  onStopList: false,
  variations: [],
  modifiers: [],
};

interface CartItem {
  id: string;
  productId: string;
  productName: string;
  quantity: number;
  unitPriceCents: number;
  unitPrepSeconds: number;
  variationIds: string[];
  modifiers: Record<string, number>;
  notes: string | null;
}

function json(route: Route, body: unknown, status = 200): Promise<void> {
  return route.fulfill({ status, contentType: 'application/json', body: JSON.stringify(body) });
}

/** Next four quarter-hour slots, the first one already full. */
function slots(): unknown[] {
  const SLOT_MS = 15 * 60 * 1000;
  const start = Math.ceil(Date.now() / SLOT_MS) * SLOT_MS;
  return Array.from({ length: 4 }, (_, i) => ({
    startsAt: new Date(start + i * SLOT_MS).toISOString(),
    endsAt: new Date(start + (i + 1) * SLOT_MS).toISOString(),
    taken: i === 0 ? 8 : 1,
    capacity: 8,
    available: i !== 0,
  }));
}

export interface FakeApi {
  /** Orders the app has created, newest last. */
  readonly orders: Array<Record<string, unknown>>;
}

/**
 * Installs the fake API on a page and returns a handle to inspect what the
 * app actually sent — asserting on the created order is how a journey test
 * proves the checkout passed the right slot and totals through.
 */
export async function installFakeApi(page: Page, opts: { pointsBalance?: number } = {}): Promise<FakeApi> {
  const items: CartItem[] = [];
  const orders: Array<Record<string, unknown>> = [];
  let nextId = 1;

  const cart = () => {
    const subtotalCents = items.reduce((sum, i) => sum + i.unitPriceCents * i.quantity, 0);
    const longestPrep = items.reduce((max, i) => Math.max(max, i.unitPrepSeconds), 0);
    return {
      id: 'cart-1',
      storeId: STORE.id,
      currency: STORE.currency,
      subtotalCents,
      // Mirrors the server: store overhead plus the longest single item.
      etaSeconds: items.length === 0 ? 0 : 420 + longestPrep,
      items: items.map((i) => ({
        id: i.id,
        productId: i.productId,
        productName: i.productName,
        quantity: i.quantity,
        unitPriceCents: i.unitPriceCents,
        totalCents: i.unitPriceCents * i.quantity,
        variationIds: i.variationIds,
        modifiers: i.modifiers,
        notes: i.notes,
      })),
    };
  };

  await page.route('**/api/**', async (route) => {
    const request = route.request();
    const url = new URL(request.url());
    const path = url.pathname.replace(/^\/api/, '');
    const method = request.method();

    // ── Auth ────────────────────────────────────────────────────────────
    if (path === '/auth/me') {
      return json(route, {
        id: 'user-1',
        phone: null,
        email: 'sam@example.com',
        name: 'Sam',
        locale: 'EN',
        currency: 'AED',
        role: 'CUSTOMER',
        telegramUserId: null,
      });
    }

    // ── Catalogue ───────────────────────────────────────────────────────
    if (path === '/stores') return json(route, [STORE]);
    if (path.match(/^\/stores\/[^/]+\/pickup-slots$/)) return json(route, slots());
    if (path.match(/^\/stores\/[^/]+\/menu$/)) {
      return json(route, {
        storeId: STORE.id,
        storeSlug: STORE.slug,
        categories: [
          {
            id: 'cat-1',
            slug: 'coffee',
            name: 'Coffee',
            description: null,
            iconUrl: null,
            sortOrder: 0,
            availableFrom: null,
            availableTo: null,
            products: [PRODUCT],
          },
        ],
      });
    }
    if (path.match(/^\/stores\/[^/]+$/)) return json(route, STORE);
    if (path.match(/^\/products\/[^/]+$/)) return json(route, PRODUCT);

    // ── Loyalty ─────────────────────────────────────────────────────────
    if (path === '/loyalty/me') {
      return json(route, {
        userId: 'user-1',
        pointsBalance: opts.pointsBalance ?? 0,
        lifetimePoints: opts.pointsBalance ?? 0,
        tier: 'SILVER',
        nextTier: 'GOLD',
        pointsToNextTier: 1500,
        tierProgressPercent: 0,
        recent: [],
      });
    }
    if (path === '/loyalty/redeem/quote') {
      const body = request.postDataJSON() as { points: number; payableCents: number };
      const affordable = Math.min(body.points, opts.pointsBalance ?? 0, body.payableCents);
      const points = affordable >= 100 ? Math.floor(affordable) : 0;
      return json(route, {
        points,
        discountCents: points,
        balance: opts.pointsBalance ?? 0,
        pointValueCents: 1,
        minPoints: 100,
      });
    }

    // ── Cart ────────────────────────────────────────────────────────────
    if (path === '/cart' && method === 'GET') return json(route, cart());
    if (path === '/cart/items' && method === 'POST') {
      const body = request.postDataJSON() as { productId: string; quantity: number };
      items.push({
        id: `item-${nextId++}`,
        productId: body.productId,
        productName: PRODUCT.name,
        quantity: body.quantity ?? 1,
        unitPriceCents: PRODUCT.basePriceCents,
        unitPrepSeconds: PRODUCT.prepTimeSeconds,
        variationIds: [],
        modifiers: {},
        notes: null,
      });
      return json(route, cart());
    }

    // ── Orders ──────────────────────────────────────────────────────────
    if (path === '/orders' && method === 'POST') {
      const body = request.postDataJSON() as Record<string, unknown>;
      const subtotalCents = cart().subtotalCents;
      const order = {
        id: 'order-1',
        orderCode: '4832',
        status: 'CREATED',
        fulfillmentType: body['fulfillmentType'] ?? 'PICKUP',
        pickupMode: body['pickupMode'] ?? 'ASAP',
        pickupAt: (body['pickupAt'] as string) ?? new Date(Date.now() + 8 * 60_000).toISOString(),
        storeId: STORE.id,
        storeName: STORE.name,
        currency: STORE.currency,
        subtotalCents,
        discountCents: 0,
        taxCents: Math.round((subtotalCents * 500) / 10_500),
        deliveryFeeCents: 0,
        giftCardCents: 0,
        totalCents: subtotalCents,
        qrToken: 'qr-token',
        etaSeconds: 480,
        items: cart().items,
        createdAt: new Date().toISOString(),
        // Echoed so the test can assert what checkout actually sent.
        _request: body,
      };
      orders.push(order);
      items.length = 0;
      return json(route, order, 201);
    }
    if (path.match(/^\/orders\/[^/]+$/) && method === 'GET') {
      return json(route, orders[orders.length - 1] ?? {}, orders.length > 0 ? 200 : 404);
    }
    if (path.startsWith('/me/orders')) return json(route, orders);

    // Anything unmapped answers empty rather than hanging the page.
    return json(route, {});
  });

  return { orders };
}

File: apps/mobile/test/helpers/fake_api.dart
import 'package:dio/dio.dart';
import 'package:mocktail/mocktail.dart';
import 'package:takeaway_api/takeaway_api.dart';

/// In-memory stand-in for the REST API: a store, a small menu, a server-side
/// cart and orders — enough for the app's flows to run end to end in widget
/// tests. Records what the app sent so tests can assert on it.
class FakeApi extends Fake implements TakeAwayApi {
  FakeApi({this.flags = const FeatureFlags()});

  FeatureFlags flags;
  final added = <AddCartItemRequest>[];
  final created = <CreateOrderRequest>[];
  final paid = <Map<String, dynamic>>[];
  final cancelled = <String>[];
  Order? currentOrder;
  List<BoundCard> boundCards = const [];

  /// Bank verdict for the next card payment; null means it goes through.
  String? declineWith;

  /// False = switched on but outside working hours, as the API reports it.
  bool storeOpenNow = true;

  /// Latency of reading the cart; the reply is the cart as it was when asked.
  Duration cartDelay = Duration.zero;

  /// Replaces the two-category menu — see [categoryJson].
  List<Map<String, dynamic>>? menuCategories;

  static const storeJson = <String, dynamic>{
    'id': 'st_1',
    'brandId': 'br_1',
    'slug': 'noname-center',
    'name': 'NoName — центр',
    'addressLine': 'ул. 25 Октября, 94',
    'city': 'Тирасполь',
    'country': 'MD',
    'latitude': 0,
    'longitude': 0,
    'status': 'OPEN',
    'fulfillmentTypes': ['TAKEAWAY'],
    'pickupPointType': 'COUNTER',
    'busyMeter': 20,
    'currentEtaSeconds': 420,
    'taxRateBps': 0,
    'taxIncludedInPrice': true,
    'currency': 'MDL',
    'heroImageUrl': null,
    'distanceMeters': null,
  };

  static Map<String, dynamic> productJson(String id, String name, int price, {String category = 'cat_coffee'}) => {
    'id': id,
    'categoryId': category,
    'slug': id,
    'name': name,
    'description': 'Двойной эспрессо и молоко.',
    'basePriceCents': price,
    'prepTimeSeconds': 180,
    'caffeineLevel': 3,
    'calories': 180,
    'proteinsGrams': null,
    'fatsGrams': null,
    'carbsGrams': null,
    'allergens': <String>[],
    'dietTags': <String>[],
    'imageUrls': <String>[],
    'sortOrder': 0,
    'onStopList': false,
  };

  static Map<String, dynamic> categoryJson(
    String id,
    String name,
    List<Map<String, dynamic>> products, {
    int sortOrder = 0,
  }) => {
    'id': id,
    'slug': id,
    'name': name,
    'description': null,
    'iconUrl': null,
    'sortOrder': sortOrder,
    'availableFrom': null,
    'availableTo': null,
    'products': products,
  };

  final _cartItems = <CartItem>[];
  var _itemSeq = 0;

  Cart _cart() => Cart(
    id: 'cart_1',
    userId: 'u1',
    storeId: 'st_1',
    subtotalCents: _cartItems.fold(0, (s, i) => s + i.unitPriceCents * i.quantity),
    etaSeconds: _cartItems.isEmpty ? 0 : 480,
    items: List.of(_cartItems),
    updatedAt: DateTime.utc(2026, 9, 22),
  );

  static ProductDetail latte() => ProductDetail.fromJson({
    ...productJson('p_latte', 'Латте', 2000),
    'brandId': 'br_1',
    'variations': [
      {
        'id': 'v_s',
        'type': 'SIZE',
        'name': 'S',
        'priceDeltaCents': 0,
        'prepTimeDeltaSeconds': 0,
        'sortOrder': 0,
        'isDefault': true,
      },
      {
        'id': 'v_m',
        'type': 'SIZE',
        'name': 'M',
        'priceDeltaCents': 500,
        'prepTimeDeltaSeconds': 20,
        'sortOrder': 1,
        'isDefault': false,
      },
      {
        'id': 'v_oat',
        'type': 'MILK',
        'name': 'Овсяное',
        'priceDeltaCents': 800,
        'prepTimeDeltaSeconds': 0,
        'sortOrder': 1,
        'isDefault': false,
      },
      {
        'id': 'v_cow',
        'type': 'MILK',
        'name': 'Коровье',
        'priceDeltaCents': 0,
        'prepTimeDeltaSeconds': 0,
        'sortOrder': 0,
        'isDefault': true,
      },
    ],
    'modifiers': [
      {
        'id': 'm_shot',
        'slug': 'shot',
        'name': 'Доп. шот',
        'priceDeltaCents': 600,
        'prepTimeDeltaSeconds': 25,
        'minCount': 0,
        'maxCount': 3,
        'sortOrder': 0,
      },
    ],
  });

  static ProductDetail croissant() => ProductDetail.fromJson({
    ...productJson('p_croissant', 'Круассан', 1500, category: 'cat_food'),
    'brandId': 'br_1',
    'variations': <Object>[],
    'modifiers': <Object>[],
  });

  void seedCart(ProductDetail product, {int quantity = 1}) {
    _cartItems.add(
      CartItem(
        id: 'ci_${++_itemSeq}',
        productId: product.id,
        productName: product.name,
        quantity: quantity,
        variationIds: const [],
        modifiers: const {},
        unitPriceCents: product.basePriceCents,
        unitPrepSeconds: product.prepTimeSeconds,
      ),
    );
  }

  static Order sampleOrder({String status = 'CREATED', Map<String, dynamic>? payment}) => Order.fromJson({
    'id': 'ord_1',
    'orderCode': '4821',
    'qrToken': 'qr_1',
    'status': status,
    'pickupMode': 'ASAP',
    'fulfillmentType': 'PICKUP',
    'pickupAt': DateTime.now().toUtc().add(const Duration(minutes: 8)).toIso8601String(),
    'subtotalCents': 2000,
    'discountCents': 0,
    'taxCents': 0,
    'totalCents': 2000,
    'currency': 'MDL',
    'storeId': 'st_1',
    'storeName': 'NoName — центр',
    'storeLatitude': 0,
    'storeLongitude': 0,
    'storeAddress': 'ул. 25 Октября, 94',
    'items': [
      {
        'id': 'it_1',
        'productSnapshot': {
          'id': 'p_croissant',
          'name': 'Круассан',
          'variationIds': <String>[],
          'modifiers': <String, int>{},
        },
        'quantity': 1,
        'unitPriceCents': 2000,
        'totalCents': 2000,
      },
    ],
    'createdAt': DateTime.now().toUtc().toIso8601String(),
    'payment': ?payment,
  });

  // ── Catalog ────────────────────────────────────────────────────────────

  @override
  Future<FeatureFlags> features() async => flags;

  @override
  Future<TelegramAuthConfig> telegramConfig() async => const TelegramAuthConfig();

  @override
  Future<List<Store>> stores({double? lat, double? lng, int? radius}) async => [
    Store.fromJson({...storeJson, 'openNow': storeOpenNow}),
  ];

  @override
  Future<StoreDetail> store(String idOrSlug) async => StoreDetail.fromJson({
    ...storeJson,
    'openNow': storeOpenNow,
    'timezone': 'Europe/Chisinau',
    'phone': null,
    'email': null,
    'minOrderCents': 0,
    'galleryUrls': <String>[],
    'workingHours': <Object>[],
    'brand': {'id': 'br_1', 'slug': 'noname', 'name': 'NoName Coffee', 'logoUrl': null, 'themeOverrides': null},
  });

  @override
  Future<Menu> menu(String idOrSlug) async => Menu.fromJson({
    'storeId': 'st_1',
    'storeSlug': 'noname-center',
    'categories':
        menuCategories ??
        [
          {
            'id': 'cat_coffee',
            'slug': 'coffee',
            'name': 'Кофе',
            'description': null,
            'iconUrl': null,
            'sortOrder': 0,
            'availableFrom': null,
            'availableTo': null,
            'products': [productJson('p_latte', 'Латте', 2000)],
          },
          {
            'id': 'cat_food',
            'slug': 'food',
            'name': 'Выпечка',
            'description': null,
            'iconUrl': null,
            'sortOrder': 1,
            'availableFrom': null,
            'availableTo': null,
            'products': [productJson('p_croissant', 'Круассан', 1500, category: 'cat_food')],
          },
        ],
  });

  @override
  Future<ProductDetail> product(String idOrSlug) async => idOrSlug == 'p_latte' ? latte() : croissant();

  /// Fixed per instance so a test can compare what it tapped with what was sent.
  late final DateTime _slotsStart = () {
    final soon = DateTime.now().toUtc().add(const Duration(minutes: 30));
    return DateTime.utc(soon.year, soon.month, soon.day, soon.hour, soon.minute - soon.minute % 15);
  }();

  @override
  Future<List<PickupSlot>> pickupSlots(String idOrSlug) async {
    final start = _slotsStart;
    return [
      for (var i = 0; i < 6; i++)
        PickupSlot(
          startsAt: start.add(Duration(minutes: 15 * i)),
          endsAt: start.add(Duration(minutes: 15 * (i + 1))),
          taken: i == 1 ? 8 : 0,
          capacity: 8,
          available: i != 1,
        ),
    ];
  }

  // ── Cart ───────────────────────────────────────────────────────────────

  @override
  Future<Cart> cart(String storeId) async {
    final snapshot = _cart();
    if (cartDelay > Duration.zero) await Future<void>.delayed(cartDelay);
    return snapshot;
  }

  @override
  Future<Cart> addCartItem(AddCartItemRequest body) async {
    added.add(body);
    final product = await this.product(body.productId);
    var price = product.basePriceCents;
    for (final v in product.variations) {
      if (body.variationIds?.contains(v.id) ?? false) price += v.priceDeltaCents;
    }
    for (final m in product.modifiers) {
      price += (body.modifiers?[m.id] ?? 0) * m.priceDeltaCents;
    }
    _cartItems.add(
      CartItem(
        id: 'ci_${++_itemSeq}',
        productId: product.id,
        productName: product.name,
        quantity: body.quantity,
        variationIds: body.variationIds ?? const [],
        modifiers: body.modifiers ?? const {},
        unitPriceCents: price,
        unitPrepSeconds: product.prepTimeSeconds,
        notes: body.notes,
      ),
    );
    return _cart();
  }

  @override
  Future<Cart> updateCartItem(String itemId, Map<String, dynamic> patch) async {
    final index = _cartItems.indexWhere((i) => i.id == itemId);
    final item = _cartItems[index];
    _cartItems[index] = CartItem(
      id: item.id,
      productId: item.productId,
      productName: item.productName,
      quantity: patch['quantity'] as int? ?? item.quantity,
      variationIds: item.variationIds,
      modifiers: item.modifiers,
      unitPriceCents: item.unitPriceCents,
      unitPrepSeconds: item.unitPrepSeconds,
    );
    return _cart();
  }

  @override
  Future<Cart> removeCartItem(String itemId) async {
    _cartItems.removeWhere((i) => i.id == itemId);
    return _cart();
  }

  @override
  Future<Cart> clearCart(String storeId) async {
    _cartItems.clear();
    return _cart();
  }

  // ── Orders ─────────────────────────────────────────────────────────────

  @override
  Future<Order> createOrder(CreateOrderRequest body) async {
    created.add(body);
    _cartItems.clear();
    return currentOrder = sampleOrder();
  }

  @override
  Future<Order> order(String id) async => currentOrder ?? sampleOrder();

  @override
  Future<Order> cancelOrder(String id) async {
    cancelled.add(id);
    return currentOrder = sampleOrder(status: 'CANCELLED');
  }

  @override
  Future<CustomerLocationResult> reportLocation(String id, Map<String, dynamic> body) async =>
      const CustomerLocationResult(level: 'HERE', distanceM: 0, recorded: true);

  @override
  Future<List<OrderSummary>> myOrders({String group = 'ALL', int take = 30}) async => const [];

  @override
  Future<LoyaltyAccount> loyalty() async => const LoyaltyAccount(
    userId: 'u1',
    pointsBalance: 0,
    lifetimePoints: 0,
    tier: LoyaltyTier.silver,
    pointsToNextTier: 500,
    tierProgressPercent: 0,
    recent: [],
    nextTier: LoyaltyTier.gold,
  );

  @override
  Future<List<BoundCard>> cards() async => boundCards;

  @override
  Future<ChargeResult> payWithCard(Map<String, dynamic> body) async {
    paid.add(body);
    final reason = declineWith;
    if (reason != null) {
      declineWith = null;
      throw DioException(
        requestOptions: RequestOptions(path: '/payments/agroprombank/pay'),
        response: Response<Object?>(
          requestOptions: RequestOptions(path: '/payments/agroprombank/pay'),
          statusCode: 400,
          data: {'statusCode': 400, 'message': reason},
        ),
        type: DioExceptionType.badResponse,
      );
    }
    return const ChargeResult(paymentId: 'pay_1', status: 'REQUIRES_ACTION', amountCents: 2000);
  }

  static BoundCard card({String id = 'card_1', bool isDefault = true}) => BoundCard.fromJson({
    'id': id,
    'maskedPan': '9104 **** **** 1234',
    'embossing': null,
    'institute': '0001',
    'instituteName': 'ЗАО «Агропромбанк»',
    'label': null,
    'isDefault': isDefault,
    'cardState': 1,
    'createdAt': '2026-09-01T10:00:00.000Z',
    'lastUsedAt': null,
  });

  @override
  Future<void> registerDevice(DeviceRegistration body) async {}
}
